src/contrib/inference/dynamichmc.jl

```diff
@@ -0,0 +1,94 @@
+###
+### DynamicHMC backend - https://github.com/tpapp/DynamicHMC.jl
+###
+struct DynamicNUTS{AD, space} <: Hamiltonian{AD} end
+
+"""
+    DynamicNUTS()
+
+Dynamic No U-Turn Sampling algorithm provided by the DynamicHMC package.
+To use it, make sure you have the DynamicHMC package installed.
+
+"""
+DynamicNUTS(args...) = DynamicNUTS{ADBackend()}(args...)
+DynamicNUTS{AD}() where AD = DynamicNUTS{AD, ()}()
+function DynamicNUTS{AD}(space::Symbol...) where AD
+    DynamicNUTS{AD, space}()
+end
+
+mutable struct DynamicNUTSState <: AbstractSamplerState
+    vi::VarInfo
+    draws::Vector
+end
+
+getspace(::Type{<:DynamicNUTS{<:Any, space}}) where {space} = space
+getspace(alg::DynamicNUTS{<:Any, space}) where {space} = space
+
+function sample_init!(
+    rng::AbstractRNG,
+    model::Model,
+    spl::Sampler{<:DynamicNUTS},
+    N::Integer;
+    kwargs...
+)
+    # Set up lp function.
+    function _lp(x)
+        value, deriv = gradient_logp(x, spl.state.vi, model, spl)
+        return ValueGradient(value, deriv)
+    end
+
+    runmodel!(model, spl.state.vi, SampleFromUniform())
+
+    if spl.selector.tag == :default
+        link!(spl.state.vi, spl)
+        runmodel!(model, spl.state.vi, spl)
+    end
+
+    spl.state.draws, _ = NUTS_init_tune_mcmc(
+        FunctionLogDensity(
+            length(spl.state.vi[spl]),
+            _lp
+        ),
+        N
+    )
+end
+
+function sample_end!(
+    rng::AbstractRNG,
+    model::Model,
+    spl::Sampler{<:DynamicNUTS},
+    N::Integer;
+    kwargs...
+)
+    runmodel!(model, spl.state.vi, SampleFromUniform())
```

Comment:
Why do we need this in `sample_end!` as well?

Reply:
We really don't, good catch.